main/src/ca/uwaterloo/flix/api/lsp/provider/HighlightProvider.scala

```diff
@@ -15,195 +15,251 @@
+  private def highlightEnumSym(uri: String, sym: Symbol.EnumSym, loc: SourceLocation)(implicit root: Root): JObject = {
+    var occurs: List[SourceLocation] = Nil
+    def add(x: SourceLocation): Unit = {
+      occurs = x :: occurs
+    }
+    def check(x: Symbol.EnumSym, loc: SourceLocation): Unit = if (x == sym) { add(loc) }
 
-        case Entity.Exp(_) => mkNotFound(uri, pos)
+    object EnumSymConsumer extends Consumer {
+      override def consumeEnum(enm: TypedAst.Enum): Unit = check(enm.sym, enm.sym.loc)
+      override def consumeType(tpe: Type): Unit = getTypeSymOccur(tpe) match {
+        case Some((sym: Symbol.EnumSym, loc)) => check(sym, loc)
+        case _ => ()
+      }
+    }
 
-        case Entity.Label(label) => highlightLabel(uri, label)
+    Visitor.visitRoot(root, EnumSymConsumer, Visitor.FileAcceptor(uri))
 
-        case Entity.FormalParam(fparam) => highlightVar(uri, fparam.bnd.sym)
+    mkHighlights(loc, occurs)
+  }
 
-        case Entity.Pattern(pat) => pat match {
-          case Pattern.Var(Binder(sym, _), _, _) => highlightVar(uri, sym)
-          case Pattern.Tag(CaseSymUse(sym, _), _, _, _) => highlightCase(uri, sym)
-          case _ => mkNotFound(uri, pos)
-        }
+  private def highlightStructFieldSym(uri: String, sym: Symbol.StructFieldSym, loc: SourceLocation)(implicit root: Root): JObject = {
+    var occurs: List[SourceLocation] = Nil
+    def add(x: SourceLocation): Unit = {
+      occurs = x :: occurs
+    }
+    def check(x: Symbol.StructFieldSym, loc: SourceLocation): Unit = if (x == sym) { add(loc) }
 
-        case Entity.Pred(pred, _) => highlightPred(uri, pred)
+    object StructFieldSymConsumer extends Consumer {
+      override def consumeStructField(field: TypedAst.StructField): Unit = check(field.sym, field.sym.loc)
+      override def consumeStructFieldSymUse(symUse: SymUse.StructFieldSymUse): Unit = check(symUse.sym, symUse.loc)
+    }
 
-        case Entity.LocalVar(sym, _) => highlightVar(uri, sym)
+    Visitor.visitRoot(root, StructFieldSymConsumer, Visitor.FileAcceptor(uri))
 
-        case Entity.Type(t) => t match {
-          case Type.Cst(tc, _) => tc match {
-            case TypeConstructor.RecordRowExtend(label) => highlightLabel(uri, label)
-            case TypeConstructor.SchemaRowExtend(pred) => highlightPred(uri, pred)
-            case TypeConstructor.Enum(sym, _) => highlightEnum(uri, sym)
-            case TypeConstructor.Effect(sym) => highlightEffect(uri, sym)
-            case _ => mkNotFound(uri, pos)
-          }
-          case Type.Var(sym, _) => highlightTypeVar(uri, sym)
-          case _ => mkNotFound(uri, pos)
-        }
+    mkHighlights(loc, occurs)
+  }
 
-        case Entity.OpUse(sym, _, _) => highlightOp(uri, sym)
+  private def highlightStructSym(uri: String, sym: Symbol.StructSym, loc: SourceLocation)(implicit root: Root): JObject = {
+    var occurs: List[SourceLocation] = Nil
+    def add(x: SourceLocation): Unit = {
```

Comment:
I don't think these small add functions -- excuse the pun -- add anything

Reply:
Yeah, no. They don't. Might as well just inline them in "check" which should also be given a better name tbh